### PR TITLE
Added add dataset, editing and save as draft cataloging functionality…

### LINKED ISSUE #2085: Allow User to Save Dataset as Draft
As an internal Magda user
I want to be able to save a dataset as draft
So that I can use my auto-generated metadata

Acceptance Criteria:
1. Allow a user to save a dataset as draft into the registry 
2. Update the dataset page to show all the fields that are available in the auto-populate metadata prototype now so it is clear which fields we have populated/user has set. Test that DGA instance does not show any of this info. Show things like Spatial extent, dataset state - draft etc.
3. Put this functionality behind a feature flag so it is not available in any production instances
4. Upon clicking Save Dataset as Draft, show a success message and the dataset URL
5. At the moment, allow this functionality for admin user only

### Dev Notes

Implementation ideas:

* this could be a enumeration of states - potentially configurable in the future to support custom workflows between dataset authrors and data custodians etc
* in search - search could index the state field and have a default filter for default search (e.g. state=published) - could specify custom parameter to bring these draft datasets out - would allow us to check permissions etc later for custom states

### PR DESCRIPTION
**Pull request by nahidakbar**
### What this PR does

Added add dataset, editing and save as draft cataloging functionality:
-   Added new aspects for publishing state and spatial coverage
-   Updated Indexer and Search API to support publishingState
-   Updated new dataset demo UI page to support adding new datasets
-   Updated dataset ui page to be able to edit fields
-   Added editor / editor type abstraction for making state management simpler for large forms.

Fixes #2085

* For future, I recommend doing an audit of the schemas. There are quite a few duplicates. I've left some things out:
  * isOpenData field and accessLevel fields - design may implicate they are doing the same things in dcat-dataset-strings
  * temporal coverage aspect and temporal field in dcat-dataset-strings
  * publisher aspect and publisher in dcat-dataset-strings
  * free text spatial field in dcat-dataset-strings and the new spatial coverage aspect
  * i think creation should be its own aspects as well as the information security objects

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: magda-registry-aspects/spatial-coverage.schema.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    "$schema": "http://json-schema.org/schema#",
+    "title": "Spatial extent defined (will try to confom as much to) in geojson.",
+    "type": "object",
+    "properties": {
+        "bbox": {
```
[the remark this answers is not in the document]
Comment (nahidakbar) [reply magda-registry-aspects/spatial-coverage.schema.json]: @AlexGilleran All geojson objects can have a bbox property that contains the bounding box of the object. That's what I went for. This gives us the flexibility to store arbitrary geojson objects underneath while still having a bounding box that contains it all. Do you still want a polygon?

File: magda-registry-aspects/publishing.schema.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "$schema": "http://json-schema.org/hyper-schema#",
+    "title": "The publishing state of a dataset",
+    "description": "Specifies the state and other metadata associated with publication of data within MAGDA system.",
+    "type": "object",
+    "properties": {
+        "state": {
```
[the remark this answers is not in the document]
Comment (nahidakbar) [reply magda-registry-aspects/publishing.schema.json]: @AlexGilleran didn't enum it on purpose. I figured states and workflow will be different for each agency. I'll add the enum for now.

File: magda-registry-aspects/spatial-coverage.schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexGilleran) [reply magda-registry-aspects/spatial-coverage.schema.json]: As long as there's some way to store the polygon - by "underneath" do you mean in another aspect?

[the remark this answers is not in the document]
Comment (nahidakbar) [reply magda-registry-aspects/spatial-coverage.schema.json]: https://tools.ietf.org/html/rfc7946#section-5

So in the future, there might be a geometry property that contains a line string or a polygon or a point etc.

**Pull request merged**